### PR TITLE
Multiple scalability improvements to controller

### PR DESCRIPTION
**Pull request by jessesuen**
1. Switch to using SharedIndexInformer to provide shared resource
   state with other controllers.
2. Use workqueue in lieue of channels which prevents work backlog,
   stale state, true concurrency, and enable future improvements
   such as rate limited retry and backoff.
3. Add ability to PATCH workflows to prevent errors related to having
   too old resource versions.
4. Pod updates no longer result in an 1:1 workflow update and instead
   adds the workflow name back to the workqueue. This allow database
   updates to be batched in a single db update.

### CODE REVIEW

File: workflow/controller/controller.go [hunk 1]
```diff
@@ -113,39 +105,107 @@ func (wfc *WorkflowController) Run(ctx context.Context) error {
...
+	wfc.podInformer = wfc.newPodInformer()
+	go wfc.wfInformer.Run(ctx.Done())
+	go wfc.podInformer.Run(ctx.Done())
 
-	// Watch Workflow objects
-	_, err = wfc.watchWorkflows(ctx)
-	if err != nil {
-		log.Errorf("Failed to register watch for Workflow resource: %v", err)
-		return err
+	// Wait for all involved caches to be synced, before processing items from the queue is started
```
Comment (shrinandj) [inline workflow/controller/controller.go]: Why is this needed? Since the informers were just created, shouldn't their caches be empty anyway?

File: workflow/controller/controller.go [hunk 2]
```diff
@@ -113,39 +105,107 @@ func (wfc *WorkflowController) Run(ctx context.Context) error {
...
+	for _, informer := range []cache.SharedIndexInformer{wfc.wfInformer, wfc.podInformer} {
+		if !cache.WaitForCacheSync(ctx.Done(), informer.HasSynced) {
+			return errors.InternalError("Timed out waiting for caches to sync")
+		}
+	}
+
+	wfc.StartStatsTicker(5 * time.Minute)
+	for i := 0; i < 4; i++ {
```
Comment (shrinandj) [inline workflow/controller/controller.go]: Does this mean there are 4 worker threads/goroutines for processing workflows? And below on line 124, there are 8 threads/goroutines for processing pods?

File: workflow/controller/controller.go [hunk 3]
```diff
@@ -113,39 +105,107 @@ func (wfc *WorkflowController) Run(ctx context.Context) error {
...
+	key, quit := wfc.podQueue.Get()
+	if quit {
+		return false
+	}
+	defer wfc.podQueue.Done(key)
+
+	obj, exists, err := wfc.podInformer.GetIndexer().GetByKey(key.(string))
+	if err != nil {
+		log.Errorf("Failed to get pod '%s' from informer index: %+v", key, err)
+		return true
```
Comment (shrinandj) [inline workflow/controller/controller.go]: What does the return value indicate? If it indicates whether the entry from the queue was correctly processed or not, shouldn't this return false?

File: workflow/controller/controller.go [hunk 4]
```diff
@@ -113,39 +105,107 @@ func (wfc *WorkflowController) Run(ctx context.Context) error {
...
+	}
+	if !exists {
+		// we can get here if the workflow updator
```
Comment (shrinandj) [inline workflow/controller/controller.go]: Incomplete comment

File: workflow/controller/operator.go [hunk 5]
```diff
@@ -82,15 +81,21 @@ func (wfc *WorkflowController) operateWorkflow(wf *wfv1.Workflow) {
 			woc.markWorkflowFailed(fmt.Sprintf("invalid spec: %s", err.Error()))
 			return
 		}
+	} else {
+		err = woc.podReconciliation()
```
Comment (shrinandj) [inline workflow/controller/operator.go]: An entry is made into the wfQueue after every pod update. And every entry in the wfQueue other than the first one when the workflow starts will query all pods in the workflow and reconcile their status. Isn't that too expensive?

File: workflow/controller/operator.go [hunk 6]
```diff
@@ -82,15 +81,21 @@ func (wfc *WorkflowController) operateWorkflow(wf *wfv1.Workflow) {
...
+		if err != nil {
+			woc.log.Errorf("%s error: %+v", wf.ObjectMeta.Name, err)
+			// TODO: we need to re-add to the workqueue, but should happen in caller
+			return
+		}
 	}
 
-	err := woc.createPVCs()
+	err = woc.createPVCs()
```
Comment (shrinandj) [inline workflow/controller/operator.go]: In a separate commit, should this code be refactored? Shouldn't the call to createPVCs only when the workflow starts?

File: workflow/controller/controller.go (hunk 1 above)
Comment (jessesuen) [reply workflow/controller/controller.go]: I think that is the problem, is that the caches should not be empty, it should be synced and represent the current state of the cluster. By proceeding without it being synced, it will have an incorrect view of the world when the control loop begins. This code snippet was taken from their examples.

File: workflow/controller/controller.go (hunk 2 above)
Comment (jessesuen) [reply workflow/controller/controller.go]: Yes, this is something that needs to be tuned, but just arbitrarily picked those for now.

File: workflow/controller/controller.go (hunk 3 above)
Comment (jessesuen) [reply workflow/controller/controller.go]: true indicates to the caller we should the control loop. A false means it should stop. This is based off their example and actual controllers

File: workflow/controller/controller.go (hunk 4 above)
Comment (jessesuen) [reply workflow/controller/controller.go]: Thanks will fix.

File: workflow/controller/operator.go (hunk 6 above)
Comment (jessesuen) [reply workflow/controller/operator.go]: If we get here, we are just about to start scheduling pods as part of the workflow. Where would this be moved?

File: workflow/controller/operator.go (hunk 5 above)
Comment (jessesuen) [reply workflow/controller/operator.go]: There are couple of things that make it not too expensive, or at least less expensive than the previous design:
1. entries to workqueues are deduped. so subsequent adds do not have any affect if the workflow key already exists in the workqueue. This is a huge improvement over the previous channel approach.
2. podReconciliation applies filters in the query to ignore both completed and pending pods as part of the workflow. This limit the working set for both the api query, and prevents us from reprocessing a pod that we don't care about anymore.
3. pod updates will only enqueue workflows upon a state change, so reconciliation should only happen when something changes. In the near future, I want to be even smarter about the state change to see if the change is even something we care about (e.g. we should not care if pod goes from Initializing to Running)
4. previously, pod updates were updating workflows one by one after every pod update. This was a very slow process, resulted in a huge number of chatter with the API server, and was causing a huge backlog in the controller processing channel/loop. By re-listing pods in podReconciliation upon pod updates, there is a significant reduction in the number of api queries and updates.

Comment (jessesuen) [reply workflow/controller/operator.go]: I should add that there is a consequence to this approach that I am still trying to work through. Because we defer examination of pod state to podReconciliation, there is now a chance for pods to become deleted from underneath the controller. I am encountering this in the pod limits testing when the controller is spending 3+ minutes just submitting pods, and by the time that the controller is done submitting, some of the pods ran to completion and were even deleted by kubernetes.

This is a consequence of using SharedInformer over regular informers. From their docs:

"SharedInformer has a shared data cache and is capable of distributing notifications for changes to the cache to multiple listeners who registered via AddEventHandler. If you use this, there is one behavior change compared to a standard Informer. When you receive a notification, the cache will be AT LEAST as fresh as the notification, but it MAY be more fresh. "

File: workflow/controller/operator.go (hunk 6 above)
Comment (shrinandj) [reply workflow/controller/operator.go]: Isn't operateWorkflow() called for every update?

runWorker -> processNextItem -> operateWorkflow?

File: workflow/controller/controller.go (hunk 1 above)
Comment (shrinandj) [reply workflow/controller/controller.go]: oh.. ok. A comment here could help understand the need to sync state.

File: workflow/controller/controller.go (hunk 2 above)
Comment (shrinandj) [reply workflow/controller/controller.go]: Let's file bugs for known improvements so that it isn't forgotten.

File: workflow/controller/controller.go (hunk 3 above)
Comment (shrinandj) [reply workflow/controller/controller.go]: IMO, this is a bit unintuitive. What do you think about moving all the code from this method into podWorker itself? The for loop could simply be an infinite loop and instead of return false, there could be a break. This would make it clear that the loop is getting terminated on some events (like quit)

Comment (jessesuen) [reply workflow/controller/controller.go]: Take a look at all the controllers in kubernetes/pkg/controller/*/*_controller.go. This is how kubernetes models all their controllers. I rather follow their lead since it will make it more understandable for people who have written other controllers.

File: workflow/controller/operator.go (hunk 5 above)
Comment (shrinandj) [reply workflow/controller/operator.go]: I see. Since each pod update is creating an entry in the workflow queue, are these other side-effects like higher memory consumption, etc. In extreme cases, this could cause the controller to crash.

This certainly seems better than the previous approach. In subsequent commits, we should consider:

* Having separate queues for submit and update (some rate limiting could be applied to workflow creation)
* As you mentioned, only put entries in the workflow queue when pods complete (success or failure)
* In the pod queue, can we store the actual status of the pod when the event happened? That way, would avoid having to go the API server when the pod entry is processed. Also, this will eliminate the problem of the pod having been gc'd when the controller processed it.

Comment (jessesuen) [reply workflow/controller/operator.go]: > are these other side-effects like higher memory consumption, etc.

There will be no increase in memory because with workqueues, work items are simple strings and are deduped. A second Add() to the queue with an existing key will not be created if the key already exists. This effectively means that the maximum number of items in the workqueue will be less than or equal to the number of workflows in the system. This is contrary to the channel approach because channels have an entry for *every* event, and cannot be not deduped.

> In the pod queue, can we store the actual status of the pod when the event happened? That way, would avoid having to go the API server when the pod entry is processed. 

It is not possible to store additional metadata/state information in workqueue because work items are simple strings. I imagine it would also make it difficult to dedup. But one benefit of using the SharedIndexInformer, it means that we can take advantage of the local cache instead of the actual API server query.

File: workflow/controller/controller.go (hunk 3 above)
Comment (shrinandj) [reply workflow/controller/controller.go]: Aah, cool!

One of them called namespace_controller (`kubernetes/pkg/controller/namespace/namespace_controller.go`)  does something similar to what was mentioned above:


* Run() creates the workers
```
        for i := 0; i < workers; i++ {
                go wait.Until(nm.worker, time.Second, stopCh)
        }
```

* The nm.worker method has all the required code within it.
```
// worker processes the queue of namespace objects.
// Each namespace can be in the queue at most once.
// The system ensures that no two workers can process
// the same namespace at the same time.
func (nm *NamespaceController) worker() {
        workFunc := func() bool {
                key, quit := nm.queue.Get()
                if quit {
                        return true
                }
                defer nm.queue.Done(key)

                // more logic here
                return false
        }

        for {
                quit := workFunc()

                if quit {
                        return
                }
        }
}
```

Since `workFunc` doesn't have it's own independent existence, it doesn't have any unintuitive behavior. This model works too.

File: workflow/controller/operator.go (hunk 6 above)
Comment (jessesuen) [reply workflow/controller/operator.go]: Yes, but createPVCs is idempotent and becomes a no-op on subsequent operateWorkflow() calls.

Comment (shrinandj) [reply workflow/controller/operator.go]: Right. It just makes it harder to read and understand.

**Comment by jessesuen** [comment]
My latest change addresses the issue of pods becoming deleted from underneath the controller, among other things. It accomplishes this by honoring a deadline for workflow operations. If the operate exceeds the deadline (currently 10s), it saves whatever work it already did, and requeues itself back to the workqueue. This has several key benefits:
1. progress for workflows with very long pod submission will now be visible to the end user
2. relinquishing control prevents starvation of other workqueue items.
3. requeuing the workflow allows pod reconciliation to take place at regular intervals. pod reconciliation enables the controller to see when pods complete much quicker, and reduces the chance that a pod can become deleted without the controller handling the event.

File: workflow/controller/controller.go (hunk 2 above)
Comment (jessesuen) [reply workflow/controller/controller.go]: Issue #568 filed.
